src/python_testing/TC_DEM_2_4.py

```diff
@@ -0,0 +1,367 @@
+        await self.send_pause_request_command(forecast.slots[0].minPauseDuration,
+                                              Clusters.DeviceEnergyManagement.Enums.AdjustmentCauseEnum.kLocalOptimization)
+        event_data = events_callback.wait_for_event_report(Clusters.DeviceEnergyManagement.Events.Paused)
+
+        self.step("13a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kPaused)
+
+        self.step("13b")
+        forecast = await self.read_dem_attribute_expect_success(attribute="Forecast")
+        asserts.assert_equal(forecast.forecastUpdateReason,
+                             Clusters.DeviceEnergyManagement.Enums.ForecastUpdateReasonEnum.kLocalOptimization)
+
+        self.step("14")
+        await self.send_resume_request_command()
+        event_data = events_callback.wait_for_event_report(Clusters.DeviceEnergyManagement.Events.Resumed)
+        asserts.assert_equal(event_data.cause, Clusters.DeviceEnergyManagement.Enums.CauseEnum.kCancelled)
+
+        self.step("14a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kOnline)
+
+        self.step("15")
+        await self.send_pause_request_command(forecast.slots[0].minPauseDuration,
+                                              Clusters.DeviceEnergyManagement.Enums.AdjustmentCauseEnum.kLocalOptimization)
+        event_data = events_callback.wait_for_event_report(Clusters.DeviceEnergyManagement.Events.Paused)
+
+        self.step("15a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kPaused)
+
+        self.step("16")
+        logging.info(f"Sleeping for forecast.slots[0].minPauseDuration {forecast.slots[0].minPauseDuration}s")
+        time.sleep(forecast.slots[0].minPauseDuration)
+
+        self.step("16a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kOnline)
+
+        self.step("17")
+        await self.send_test_event_trigger_pausable_next_slot()
+
+        self.step("17a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kOnline)
+
+        self.step("17b")
+        forecast = await self.read_dem_attribute_expect_success(attribute="Forecast")
+        print(forecast)
+        asserts.assert_equal(forecast.activeSlotNumber, 1)
+
+        self.step("18")
+        await self.send_pause_request_command(forecast.slots[0].minPauseDuration,
+                                              Clusters.DeviceEnergyManagement.Enums.AdjustmentCauseEnum.kLocalOptimization,
+                                              expected_status=Status.ConstraintError)
+
+        self.step("18a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kOnline)
+
+        self.step("19")
+        await self.send_resume_request_command(expected_status=Status.Failure)
```

Comment:
I think this should be InvalidInState error (or test plan is wrong?)

Reply:
It should be InvalidInState - code + test changed